### DISCUSSION TITLE
**"Variable is optimized away and not available."**, how can I view those variables in vscode?

### DISCUSSION THREAD

**Discussion by xiongmao86** [rust-lang/rust-analyzer#14706] [Q&A] [answered]
Hi, I am trying to debug a project call `typst`, when I debug the main function, in the `local`'s view, some of the variables are not visible, and showing `Variable is optimized away and not available. 

is that because of the way I build? I build with `cargo build`, and the output is `dev [unoptimized + debuginfo]`

P.S. Thanks for making this great app, I can't wait to learn more about it.

**Comment by lnicola** [comment, marked as answer]
Hi. This sometimes happens, e.g. on unused variables. Regardless of that, I don't think you can do much about it. Native language debuggers will always give a worse experience than those for managed languages.

We don't do anything besides setting up the debugging extension. Depending on your platform, you can try another one, there's at least three: C/C++, Native Debug and CodeLLDB.

**Reply by xiongmao86** [reply]
Hi, @lnicola, thank you for your help. English is not my first language, and I am a little bit confused, the c++ is the c++ extension of microsoft, I have heard of, what is Native Debug refer to?

**Reply by lnicola** [reply]
- https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools
 - https://marketplace.visualstudio.com/items?itemName=webfreak.debug
 - https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb

**Reply by xiongmao86** [reply]
Thank you very much. @lnicola.

**Reply by lnicola** [reply]
If you try them, can you post your results and operating system, toolchain (`-msvc` or `-gnu` if Windows etc.), maybe a small snippet showing the problem for anyone else who has this problem?

**Reply by xiongmao86** [reply]
I experiment a little bit. The description would be incorrect, sometimes I have to step one more statement for variables of the current statement to show up, plus the description saying it's optimized away, so I assume they are not available. But they are available, only one step away.

**Comment by Universal-Invariant** [comment]
It seems that variable is optimized away includes variables that have no yet actually been created. Ideally such variables would not be displayed until they are actually valid.

**Reply by lnicola** [reply]
You're right, but this has been my experience with pretty much all native code debuggers (though the MSVC one is pretty friendly if you enable that Just My Code option), and we can't do anything about it.

**Reply by Universal-Invariant** [reply]
But simply showing a useless msg only clutters up the list. There is only so much room for variables without having to scroll so no need to show that info to the user as it doesn't inform him of anything meaningful except that there is or was a variable with that name. All such variables could be either not displayed at all until they are meaningful or put in a non-expanded sub-list.

Also, at least with cppvsdbg which I had to switch to from codelldb since it fails to work in any useful way, any use of "F5" to continue will result in most variables showing "variable is optimized away" until a single step is done in which then they are populated with values. This is a bit annoying.  

I don't think this is anything specific to do with rust-analyzer and just likely the debuggers not being tailored to rust specifically.  Clearly there is some code somewhere that displays the variables with the text and that should be nuked for rust when the text is "Variable is optimized away".

**Reply by lnicola** [reply]
> debuggers not being tailored to rust specifically

The UX is exactly the same in C and C++ :sob:.

**Reply by Universal-Invariant** [reply]
well, then I don't know what to say. Seems likes someone dropped the ball on that one. Not sure why people think it would be acceptable to pollute the lists with junk.

**Comment by Universal-Invariant** [comment]
Ok, after playing around with cppvsdbg and having strange issues(works about 90% of the time) I found out that codelldb requires a special gnu version for windows. E.g., codelldb requires rust-<ver>-x86_64-pc-windows-gnu.msi.

I installed rustup from the link from main download page(https://www.rust-lang.org/tools/install) when one needs to go https://forge.rust-lang.org/infra/other-installation-methods.html and download the windows gnu version.

I think I got the info that one needs or can use a different arch:: https://stackoverflow.com/questions/48067874/how-do-i-inspect-local-variables-or-use-the-watch-window-with-rust-code-using-vs but I also found other people saying similar things about issues. I think maybe codelldb should try to determine if the gnu version is installed and warn the user if it is not.


`
{
	// Use IntelliSense to learn about possible attributes.
	// Hover to view descriptions of existing attributes.
	// For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
	"version": "0.2.0",
	"configurations": [
		{
			"name": "Debug2",
			"type": "lldb",
			"request": "launch",
			"program": "${workspaceFolder}/target/debug/${workspaceFolderBasename}.exe",
			"preLaunchTask": "rust: cargo build",
			"console": "internalConsole",
			"args": [],
			"cwd": "${cwd}",
			"presentation": {
				"hidden": false,
				"group": "",
				"order": 1,
				"reveal": "never",
				"focus": false,
				"panel": "shared",
				"showReuseMessage": false,
				"clear": false
			},
		},
		{
			"name": "Debug",
			"type": "cppvsdbg",
			"request": "launch",
			"program": "${workspaceFolder}/target/debug/${workspaceFolderBasename}.exe",
			"preLaunchTask": "rust: cargo build",
			"stopAtEntry": false,
			"environment": [],
			"console": "internalConsole",
			"args": [],
			"cwd": "${cwd}",
			"presentation": {
				"hidden": false,
				"group": "",
				"order": 1,
				"reveal": "never",
				"focus": false,
				"panel": "shared",
				"showReuseMessage": false,
				"clear": false
			},
		},
	]
}
`


`
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "cargo",
			"command": "build",
			//"args":["+stable-msvc"],
			"problemMatcher": [
				"$rustc"
			],
			"group": "build",
			"label": "build",
			"presentation": {
				"echo": true,
				"reveal": "never",
				"focus": false,
				"panel": "shared",
				"showReuseMessage": true,
				"clear": false
			}
		},
        {
			"group": "build",
            "label": "clean",
            "type": "shell",
            "command": "Remove-Item",
			"args":["-Path", "\"${workspaceFolder}\\target\\debug\\${workspaceFolderBasename}.exe\"", "-Force"],
            "problemMatcher": []
        },
        {
            "label": "rust: cargo build",
            "dependsOn": ["clean", "build"],
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }
	]

}`

Not saying all of the above is necessary or makes sense, just how I got codelldb to work correctly when I wrote it off at first and only after I started having problems with cppvsdbg not working well.
